mismi-s3/src/Mismi/S3/Commands.hs

```diff
@@ -255,21 +256,34 @@ listObjects :: Address -> AWS ([Address], [Address])
 listObjects a =
   (\(p, k) -> (Address (bucket a) <$> p, Address (bucket a) <$> k) )<$> getObjects a
 
--- list the addresses, keys first, then prefixes
 list :: Address -> AWS [Address]
 list a =
-  (\(p, k) -> k <> p) <$> listObjects a
+  list' a >>= ($$ DC.consume)
+
+list' :: Address -> AWS (Source AWS Address)
+list' a@(Address (Bucket b) (Key k)) = do
+  let pp kk = if T.null kk then "" else if T.isSuffixOf "/" kk then kk else kk <> "/"
+  e <- ask
+  pure . hoist (retryConduit e) $ (paginate $ AWS.listObjects (BucketName b) & loPrefix .~ Just (pp k) & loDelimiter .~ Just '/') =$= liftAddressAndPrefix a
+
+liftAddressAndPrefix :: Address -> Conduit ListObjectsResponse AWS Address
+liftAddressAndPrefix a =
+  DC.mapFoldable (\r ->
+       fmap (\o -> let ObjectKey t = o ^. oKey in a { key = Key t })(r ^. lorsContents)
+    <> join (traverse (\cp -> maybeToList .fmap (\cp' -> a { key = Key cp' }) $ cp ^. cpPrefix) (r ^. lorsCommonPrefixes))
+  )
+
 
 download :: Address -> FilePath -> AWS ()
 download = downloadWithMode Fail
 
 downloadWithMode :: WriteMode -> Address -> FilePath -> AWS ()
 downloadWithMode mode a f = do
-  when (mode == Fail) . whenM (liftIO $ doesFileExist f) . fail $ "Can not download to a target that already exists [" <> f <> "]."
-  unlessM (exists a) . fail $ "Can not download when the source does not exist [" <> (T.unpack $ addressToText a) <> "]."
+  when (mode == Fail) . whenM (liftIO $ doesFileExist f) . throwM $ DestinationFileExists f
+  unlessM (exists a) . throwM $ SourceMissing DownloadError a
   liftIO $ createDirectoryIfMissing True (dropFileName f)
   r <- getObject' a
-  r' <- maybe (fail "shit") pure r
+  r' <- maybe (throwM $ NoContent a) pure r
```

Comment:
What if we just call `getObject'` omitting `exists` and depending on the result we can either `throwM SourceMissing` or return the result with `to bodyResponse`. This way we don't need `NoContent` exception?